Sema: Move the availability macros cache to the ASTContext

xymus opened the pull request:
The availability macros definitions are parsed from the command line and stored in a cache. The cache was in the Parser, which would have it be computed for each file using availability attributes. Let's move it to the ASTContext instead where it can generally be computed once per invocation and used across the module, speeding up the compilation process.

Also optimize how we register the buffer holding the macro definitions. Registering all of them before starting to parse their content avoids sorting them repeatedly which caused a noticeable slowdown on high usage of availability macros.

rdar://134797088

xymus committed:
Sema: Move the availability macros cache to the ASTContext

The availability macros definitions are parsed from the command line and
stored in a cache. The cache was in the Parser, which would have it be
computed for each file using availability macros. Let's move it to the
ASTContext instead where it can generally be computed once per invocation
and used across the module.

rdar://134797088

xymus commented:
@swift-ci Please smoke test

DougGregor commented on include/swift/AST/AvailabilitySpec.h:
```diff
@@ -215,6 +215,19 @@ class OtherPlatformAvailabilitySpec : public AvailabilitySpec {
+                         SmallVector<AvailabilitySpec *, 4>> VersionEntry;
+
+  bool WasParsed = false;
+  llvm::DenseMap<StringRef, VersionEntry> Impl;
```
Should this be a StringMap rather than depend on the StringRefs living forever?

xymus replied:
I think that's safe currently. The StringRefs should point to a buffer attached to the SourceManager of the ASTContext, so both the cache and the sources should be freed together. Does this sound right?

xymus replied:
Currently one of these buffers is allocated for each `-define-availability` argument. I suspect this doesn't help with the general performances of the `findBuffer` services on wide uses of availability macros, we could look into deregistering these small buffers after parsing and copy all these strings.

DougGregor commented on lib/AST/ASTContext.cpp:
```diff
@@ -426,6 +426,9 @@ struct ASTContext::Implementation {
   swift::namelookup::ImportCache TheImportCache;
 
+  /// Cache of availability macros parsed from the command line.
+  AvailabilityMacroMap TheAvailabilityMacroCache;
```
Could we use the request-evaluator here rather than add the cache directly to the ASTContext?

xymus replied:
Other requests seem to use the ASTContext itself. I'll see to use that.

xymus committed:
Parser: Register all availability macro definition buffers before parsing

The availability macro definitions from the command line are copied into
source buffers for parsing. Let's register all of these buffers in one go,
before starting to parse any of them. Parsing triggers using `findBuffer`
services which sorts all buffers. The new logic should ensure this
sorting happens only once for availability macros.

xymus commented:
@swift-ci Please smoke test

xymus commented:
Merging with 2 different optimizations. I'll put up a PR to move the cache to a request next as this optimization is time sensitive.
